src/serac/physics/utilities/functional/functional.hpp

```diff
@@ -357,20 +388,246 @@ class Functional<test(trial), execution_policy> : public mfem::Operator {
+            }
+          }
+        }
+
+      }
+
+      // mfem doesn't implement GetDofMap for some of its Nedelec elements (??),
+      // so we have to temporarily disable boundary terms for Hcurl until they do
+      if (form.boundary_integrals_.size() > 0) {
+        bool on_boundary = true;
+
+        for (int b = 0; b < num_boundary_elements; b++) {
+          form.test_space_->GetBdrElementDofs(b, test_dofs);
+          form.trial_space_->GetBdrElementDofs(b, trial_dofs);
+
+          if constexpr (test::family != Family::HCURL) {
+            const mfem::Array<int> & test_native_to_lexicographic = dynamic_cast<const mfem::TensorBasisElement *>(form.test_space_->GetBE(0))->GetDofMap();
+            ::detail::apply_permutation(test_dofs, test_native_to_lexicographic);
+          }
+
+          if constexpr (trial::family != Family::HCURL) {
+            const mfem::Array<int> & trial_native_to_lexicographic = dynamic_cast<const mfem::TensorBasisElement *>(form.trial_space_->GetBE(0))->GetDofMap();
+            ::detail::apply_permutation(trial_dofs, trial_native_to_lexicographic);
+          }
+
+          for (int i = 0; i < dofs_per_test_boundary_element; i++) {
+            for (int j = 0; j < dofs_per_trial_boundary_element; j++) {
+              for (int k = 0; k < test_vdim; k++) {
+                int test_vdof = form.test_space_->DofToVDof(::detail::get_index(test_dofs[i]), k);
+                for (int l = 0; l < trial_vdim; l++) {
+                  int trial_vdof = form.trial_space_->DofToVDof(::detail::get_index(trial_dofs[j]), l);
+                  infos.push_back({test_vdof, trial_vdof, i + dofs_per_test_boundary_element * k, j + dofs_per_trial_boundary_element * l, b, ::detail::get_sign(test_dofs[i]) * ::detail::get_sign(trial_dofs[j]), on_boundary});
+                }
+              }
+            }
+          }
+        }
+
+      }
+
+      std::sort(infos.begin(), infos.end());
+
+      int nrows = form.test_space_->GetNDofs() * form.test_space_->GetVDim();
+      row_ptr.resize(nrows + 1);
+      std::vector < ::detail::signed_index > nonzero_ids(infos.size());
+
+      int nnz = 0;
+      row_ptr[0] = 0;
+      col_ind.push_back(infos[0].global_col);
+      nonzero_ids[0] = {0, infos[0].sign};
+
+      for (size_t i = 1; i < infos.size(); i++) {
+        // increment the nonzero count every time we find a new (i,j) pair
+        nnz += (infos[i-1] != infos[i]);
+
+        nonzero_ids[i] = {nnz, infos[i].sign};
+
+        if (infos[i-1] != infos[i]) {
+          col_ind.push_back(infos[i].global_col);
+        }
+
+        for (int j = infos[i-1].global_row; j < infos[i].global_row; j++) {
+          row_ptr[j+1] = nonzero_ids[i];
+        }
+      }
+
+      ++nnz;
+      for (int j = infos.back().global_row; j < nrows; j++) {
+        row_ptr[j+1] = nnz;
+      }
+
+      element_nonzero_LUT = Array3D<::detail::signed_index>(num_elements, dofs_per_test_element * test_vdim, dofs_per_trial_element * trial_vdim);
```

Comment:
What does `LUT` mean? Comments would be appreciated.

Reply:
LUT: "look-up table"

added more comments (in-line and Doxygen) to help clarify steps and intent